applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py

```diff
@@ -102,50 +102,57 @@ class GeoMechanicsAnalysis(GeoMechanicsAnalysisBase):
         self.max_delta_time      = self.delta_time * self.max_delta_time_factor
-        self.max_iterations      = project_parameters["solver_settings"]["max_iterations"].GetInt()
-        self.min_iterations      = project_parameters["solver_settings"]["min_iterations"].GetInt()
         self.number_cycles       = project_parameters["solver_settings"]["number_cycles"].GetInt()
+
+        self.max_iterations      = project_parameters["solver_settings"]["max_iterations"].GetInt()
         self.solution_type       = project_parameters["solver_settings"]["solution_type"].GetString()
         self.reset_displacements = project_parameters["solver_settings"]["reset_displacements"].GetBool()
-        self.start_time          = project_parameters["solver_settings"]["start_time"].GetDouble()
-        self.end_time            = project_parameters["problem_data"]["end_time"].GetDouble()
         self.rebuild_level       = project_parameters["solver_settings"]["rebuild_level"].GetInt()
 
     def FinalizeSolutionStep(self):
         super().FinalizeSolutionStep()
 
-        if(self._GetSolver().GetComputingModelPart().ProcessInfo[KratosMultiphysics.NL_ITERATION_NUMBER] > self.max_iterations):
+        if (self._GetSolver().GetComputingModelPart().ProcessInfo[KratosMultiphysics.NL_ITERATION_NUMBER] > self.max_iterations):
             raise Exception("max_number_of_iterations_exceeded")
 
     def RunSolutionLoop(self):
         """This function executes the solution loop of the AnalysisStage
         It can be overridden by derived classes
         """
+
+        # store total displacement field for reset_displacements
         if self._GetSolver().settings["reset_displacements"].GetBool():
             old_total_displacements = [node.GetSolutionStepValue(KratosGeo.TOTAL_DISPLACEMENT)
                                        for node in self._GetSolver().GetComputingModelPart().Nodes]
 
         self._GetSolver().solver.SetRebuildLevel(self.rebuild_level)
 
         while self.KeepAdvancingSolutionLoop():
+            # check against max_delta_time should only be necessary when trying to scale up
             if (self.delta_time > self.max_delta_time):
                 self.delta_time = self.max_delta_time
                 KratosMultiphysics.Logger.PrintInfo(self._GetSimulationName(), "reducing delta_time to max_delta_time: ", self.max_delta_time)
-            t = self._GetSolver().GetComputingModelPart().ProcessInfo[KratosMultiphysics.TIME]
-            new_time = t + self.delta_time
-            if (new_time > self.end_time):
-                new_time = self.end_time
-                self.delta_time = new_time - t
+
+            # maximize delta_time to avoid exceeding the end_time
+            t               = self._GetSolver().GetComputingModelPart().ProcessInfo[KratosMultiphysics.TIME]
+            self.delta_time = min( self.delta_time, self.end_time - t )
```

Comment:
The logic of this section could be compressed into a single line. i.e 
`self.delta_time = min(self.delta_time, self.end_time -t, self.max_delta_time)`
However I see the purpose for logging purposes.

Reply:
Indeed